### PR TITLE
github-action: enable provenance

### PR DESCRIPTION
**Pull request by v1v**
## What does this pull request do?

Use https://github.com/github-early-access/generate-build-provenance to create build attestations for binaries and docker images

https://github.com/elastic/apm-agent-python/attestations is the one with the different executions.

### CODE REVIEW

File: .github/workflows/packages.yml [hunk 1]
```diff
@@ -28,12 +28,8 @@ jobs:
...
         with:
           python-version: "3.10"
-      - name: Install wheel
-        run: pip install --user wheel
-      - name: Building universal wheel
-        run: python setup.py bdist_wheel
-      - name: Building source distribution
-        run: python setup.py sdist
+      - name: Build packages
+        run: ./dev-utils/make-packages.sh
```
Comment (v1v) [inline .github/workflows/packages.yml:32]: This should help with reusing the script somewhere else and do some post-build steps

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -86,6 +130,9 @@ jobs:
     needs:
       - build-distribution
     runs-on: ubuntu-latest
+    permissions:
```
Comment (v1v) [inline .github/workflows/release.yml:133]: For the build attestation

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -107,25 +154,30 @@ jobs:
           if [ "${{ startsWith(github.ref, 'refs/tags') }}" == "false" ] ; then
             # for testing purposes
             echo "tag=test" >> "${GITHUB_OUTPUT}"
+            echo "latest=test-latest" >> "${GITHUB_OUTPUT}"
```
Comment (v1v) [inline .github/workflows/release.yml:157]: To help with creating tags on the fly and use them even for testing the release in a feature branch

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -107,25 +154,30 @@ jobs:
...
-      - name: Docker retag
-        run: >-
-          docker tag
-          ${{ env.DOCKER_IMAGE_NAME }}:${{ steps.setup-docker.outputs.tag }}
-          ${{ env.DOCKER_IMAGE_NAME }}:latest
-      - name: Docker push
-        if: startsWith(github.ref, 'refs/tags')
-        run: |-
-          docker push --all-tags ${{ env.DOCKER_IMAGE_NAME }}
+      - name: Build and push image
```
Comment (v1v) [inline .github/workflows/release.yml:163]: Required for the attestation

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -63,7 +107,7 @@ jobs:
           secrets: |
             secret/observability-team/ci/service-account/apm-agent-python access_key_id | AWS_ACCESS_KEY_ID ;
             secret/observability-team/ci/service-account/apm-agent-python secret_access_key | AWS_SECRET_ACCESS_KEY
-      - uses: actions/download-artifact@v3
+      - uses: actions/download-artifact@v4
```
Comment (v1v) [inline .github/workflows/release.yml:110]: Faster builds

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -6,6 +6,7 @@ on:
       - "v*.*.*"
     branches:
       - main
+      - feature/sbom
```
Comment (v1v) [inline .github/workflows/release.yml:9]: Unrequired

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -18,7 +19,22 @@ jobs:
...
 
   packages:
-    uses: ./.github/workflows/packages.yml
+    permissions:
+      id-token: write
+      contents: write
+    runs-on: ubuntu-latest
+    # Override the version if there is no tag release.
+    env:
+      ELASTIC_CI_POST_VERSION: ${{ startsWith(github.ref, 'refs/tags') && '' || github.run_id }}
```
Comment (v1v) [inline .github/workflows/release.yml:28]:
```suggestion
```

**Review by reakaleek — APPROVED** [review]

**Review by xrmx — APPROVED** [review]

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -107,25 +126,30 @@ jobs:
...
+          tags: |
+            ${{ env.DOCKER_IMAGE_NAME }}:${{ steps.setup-docker.outputs.tag }}
+            ${{ env.DOCKER_IMAGE_NAME }}:${{ steps.setup-docker.outputs.latest }}
+          build-args: |
+            AGENT_DIR=./build/dist/package/python
+
+      - name: Attest image
+        uses: github-early-access/generate-build-provenance@main
+        with:
+          subject-name: "${{ env.DOCKER_IMAGE_NAME }}:${{ steps.setup-docker.outputs.tag }}"
```
Comment (trentm) [inline .github/workflows/release.yml:150]: https://github.com/github-early-access/generate-build-provenance/tree/main?tab=readme-ov-file#container-image says:

> ... Do NOT include a tag as part of the image name -- the specific image being attested is identified by the supplied digest. ...

Comment (v1v) [reply .github/workflows/release.yml:150]: Good catch!!  Fixing it now

Comment (v1v) [reply .github/workflows/release.yml:150]: Interestingly, it works as is:

https://github.com/elastic/apm-agent-python/attestations/739913

**Pull request merged**